JsonDecodingException when decoding playlist with id 0nXKVLwZ7yk7hkU6OWsFcw

When Kotify try to decode https://open.spotify.com/playlist/0nXKVLwZ7yk7hkU6OWsFcw?si=4eddfe0f414d4550 throw a KotifyException with the following log:

```log
Caused by: tv.blademaker.kotify.exceptions.KotifyException: Unexpected JSON token at offset 338133: Expected string literal but 'null' literal was found.
Use 'coerceInputValues = true' in 'Json {}` builder to coerce nulls to default values.
JSON input: ..... : {
          "album_type" : null,
          "artists" : [ .....
        at tv.blademaker.kotify.request.Request$execute$2.invokeSuspend(Request.kt:77)
        at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
        at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
        at kotlinx.coroutines.EventLoopImplBase.processNextEvent(EventLoop.common.kt:279)
        at kotlinx.coroutines.BlockingCoroutine.joinBlocking(Builders.kt:85)
        at kotlinx.coroutines.BuildersKt__BuildersKt.runBlocking(Builders.kt:59)
        at kotlinx.coroutines.BuildersKt.runBlocking(Unknown Source)
        at kotlinx.coroutines.BuildersKt__BuildersKt.runBlocking$default(Builders.kt:38)
        at kotlinx.coroutines.BuildersKt.runBlocking$default(Unknown Source)
        at net.hugebot.audio.sources.spotify.providers.SpotifyPlaylistProvider.load(SpotifyPlaylistProvider.kt:22)
        at net.hugebot.audio.sources.spotify.SpotifyAudioSourceManager.loadItemOnce(SpotifyAudioSourceManager.kt:59)
        at net.hugebot.audio.sources.spotify.SpotifyAudioSourceManager.loadItem(SpotifyAudioSourceManager.kt:44)
        at com.sedmelluq.discord.lavaplayer.player.DefaultAudioPlayerManager.checkSourcesForItemOnce(DefaultAudioPlayerManager.java:436)
        at com.sedmelluq.discord.lavaplayer.player.DefaultAudioPlayerManager.checkSourcesForItem(DefaultAudioPlayerManager.java:418)
        at com.sedmelluq.discord.lavaplayer.player.DefaultAudioPlayerManager.lambda$createItemLoader$0(DefaultAudioPlayerManager.java:217)
        ... 6 common frames omitted
Caused by: kotlinx.serialization.json.internal.JsonDecodingException: Unexpected JSON token at offset 338133: Expected string literal but 'null' literal was found.
Use 'coerceInputValues = true' in 'Json {}` builder to coerce nulls to default values.
JSON input: ..... : {
          "album_type" : null,
          "artists" : [ .....
        at kotlinx.serialization.json.internal.JsonExceptionsKt.JsonDecodingException(JsonExceptions.kt:24)
        at kotlinx.serialization.json.internal.JsonExceptionsKt.JsonDecodingException(JsonExceptions.kt:32)
        at kotlinx.serialization.json.internal.AbstractJsonLexer.fail(AbstractJsonLexer.kt:524)
        at kotlinx.serialization.json.internal.AbstractJsonLexer.unexpectedToken(AbstractJsonLexer.kt:202)
        at kotlinx.serialization.json.internal.StringJsonLexer.consumeNextToken(StringJsonLexer.kt:74)
        at kotlinx.serialization.json.internal.StringJsonLexer.consumeKeyString(StringJsonLexer.kt:85)
        at kotlinx.serialization.json.internal.AbstractJsonLexer.consumeString(AbstractJsonLexer.kt:310)
        at kotlinx.serialization.json.internal.StreamingJsonDecoder.decodeString(StreamingJsonDecoder.kt:256)
        at kotlinx.serialization.encoding.AbstractDecoder.decodeStringElement(AbstractDecoder.kt:58)
        at tv.blademaker.kotify.models.PartialAlbum$$serializer.deserialize(Album.kt:24)
        at tv.blademaker.kotify.models.PartialAlbum$$serializer.deserialize(Album.kt:24)
        at kotlinx.serialization.json.internal.PolymorphicKt.decodeSerializableValuePolymorphic(Polymorphic.kt:59)
        at kotlinx.serialization.json.internal.StreamingJsonDecoder.decodeSerializableValue(StreamingJsonDecoder.kt:36)
        at kotlinx.serialization.encoding.AbstractDecoder.decodeSerializableValue(AbstractDecoder.kt:43)
        at kotlinx.serialization.encoding.AbstractDecoder.decodeNullableSerializableElement(AbstractDecoder.kt:79)
        at tv.blademaker.kotify.models.Track$$serializer.deserialize(Track.kt:6)
        at tv.blademaker.kotify.models.Track$$serializer.deserialize(Track.kt:6)
        at kotlinx.serialization.json.internal.PolymorphicKt.decodeSerializableValuePolymorphic(Polymorphic.kt:59)
        at kotlinx.serialization.json.internal.StreamingJsonDecoder.decodeSerializableValue(StreamingJsonDecoder.kt:36)
        at kotlinx.serialization.encoding.AbstractDecoder.decodeSerializableValue(AbstractDecoder.kt:43)
        at kotlinx.serialization.encoding.AbstractDecoder.decodeSerializableElement(AbstractDecoder.kt:70)
        at tv.blademaker.kotify.models.PlaylistPagination$Item$$serializer.deserialize(PlaylistPagination.kt:19)
        at tv.blademaker.kotify.models.PlaylistPagination$Item$$serializer.deserialize(PlaylistPagination.kt:19)
        at kotlinx.serialization.json.internal.PolymorphicKt.decodeSerializableValuePolymorphic(Polymorphic.kt:59)
        at kotlinx.serialization.json.internal.StreamingJsonDecoder.decodeSerializableValue(StreamingJsonDecoder.kt:36)
        at kotlinx.serialization.encoding.AbstractDecoder.decodeSerializableValue(AbstractDecoder.kt:43)
        at kotlinx.serialization.encoding.AbstractDecoder.decodeSerializableElement(AbstractDecoder.kt:70)
        at kotlinx.serialization.encoding.CompositeDecoder$DefaultImpls.decodeSerializableElement$default(Decoding.kt:535)
        at kotlinx.serialization.internal.ListLikeSerializer.readElement(CollectionSerializers.kt:80)
        at kotlinx.serialization.internal.AbstractCollectionSerializer.readElement$default(CollectionSerializers.kt:51)
        at kotlinx.serialization.internal.AbstractCollectionSerializer.merge(CollectionSerializers.kt:36)
        at kotlinx.serialization.internal.AbstractCollectionSerializer.deserialize(CollectionSerializers.kt:43)
        at kotlinx.serialization.json.internal.PolymorphicKt.decodeSerializableValuePolymorphic(Polymorphic.kt:59)
        at kotlinx.serialization.json.internal.StreamingJsonDecoder.decodeSerializableValue(StreamingJsonDecoder.kt:36)
        at kotlinx.serialization.encoding.AbstractDecoder.decodeSerializableValue(AbstractDecoder.kt:43)
        at kotlinx.serialization.encoding.AbstractDecoder.decodeSerializableElement(AbstractDecoder.kt:70)
        at tv.blademaker.kotify.models.PlaylistPagination$$serializer.deserialize(PlaylistPagination.kt:7)
        at tv.blademaker.kotify.models.PlaylistPagination$$serializer.deserialize(PlaylistPagination.kt:7)
        at kotlinx.serialization.json.internal.PolymorphicKt.decodeSerializableValuePolymorphic(Polymorphic.kt:59)
        at kotlinx.serialization.json.internal.StreamingJsonDecoder.decodeSerializableValue(StreamingJsonDecoder.kt:36)
        at kotlinx.serialization.encoding.AbstractDecoder.decodeSerializableValue(AbstractDecoder.kt:43)
        at kotlinx.serialization.encoding.AbstractDecoder.decodeSerializableElement(AbstractDecoder.kt:70)
        at tv.blademaker.kotify.models.Playlist$$serializer.deserialize(Playlist.kt:6)
        at tv.blademaker.kotify.models.Playlist$$serializer.deserialize(Playlist.kt:6)
        at kotlinx.serialization.json.internal.PolymorphicKt.decodeSerializableValuePolymorphic(Polymorphic.kt:59)
        at kotlinx.serialization.json.internal.StreamingJsonDecoder.decodeSerializableValue(StreamingJsonDecoder.kt:36)
        at kotlinx.serialization.json.Json.decodeFromString(Json.kt:100)
        at tv.blademaker.kotify.request.Request$execute$2.invokeSuspend(Request.kt:56)
        ... 20 common frames omitted
```

## External urls
- [Sentry Exception](https://sentry.io/share/issue/6d9434ef25494ea59000ff722a502cee/)

--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -10,7 +10,7 @@ plugins {
 }
 
 group = "tv.blademaker"
-val versionObj = Version(0, 4, 4)
+val versionObj = Version(0, 4, 5)
 version = versionObj.toString()
 
 repositories {
@@ -38,7 +38,13 @@ val dokkaOutputDir = "$buildDir/dokka"
 
 tasks {
     withType<org.jetbrains.kotlin.gradle.tasks.KotlinCompile> {
-        this.kotlinOptions.jvmTarget = "11"
+        this.kotlinOptions {
+            jvmTarget = "11"
+            freeCompilerArgs = listOf(
+                "-Xopt-in=kotlin.RequiresOptIn"
+            )
+        }
+
     }
 
     getByName<org.jetbrains.dokka.gradle.DokkaTask>("dokkaHtml") {

--- a/src/main/kotlin/tv/blademaker/kotify/Kotify.kt
+++ b/src/main/kotlin/tv/blademaker/kotify/Kotify.kt
@@ -53,10 +53,7 @@ class Kotify(
     internal val httpClient = HttpClient(CIO) {
         expectSuccess = true
         install(ContentNegotiation) {
-            json(Json {
-                isLenient = true
-                ignoreUnknownKeys = true
-            })
+            json(JSON)
         }
     }
 
@@ -119,6 +116,11 @@ class Kotify(
         internal var baseUrl: String = "https://api.spotify.com"
 
         const val VERSION = "0.4.2"
+
+        val JSON = Json {
+            isLenient = true
+            ignoreUnknownKeys = true
+        }
     }
 
     enum class Scope(

--- a/src/main/kotlin/tv/blademaker/kotify/models/Album.kt
+++ b/src/main/kotlin/tv/blademaker/kotify/models/Album.kt
@@ -15,7 +15,7 @@ enum class AlbumType {
 data class Album(
     val id: String,
     val name: String,
-    @SerialName("album_type") val albumType: String,
+    @SerialName("album_type") val albumType: AlbumType? = null,
     @SerialName("total_tracks") val totalTracks: Int,
     @SerialName("available_markets") val availableMarkets: List<String>? = null,
     @SerialName("external_urls") val externalUrls: Map<String, String>,
@@ -34,7 +34,7 @@ data class PartialAlbum(
     val id: String,
     val name: String,
     val uri: String,
-    @SerialName("album_type") val albumType: String,
+    @SerialName("album_type") val albumType: AlbumType? = null,
     @SerialName("total_tracks") val totalTracks: Int,
     @SerialName("available_markets") val availableMarkets: List<String>? = null,
     @SerialName("external_urls") val externalUrls: Map<String, String>,

--- a/src/main/kotlin/tv/blademaker/kotify/models/Album.kt
+++ b/src/main/kotlin/tv/blademaker/kotify/models/Album.kt
@@ -15,7 +15,7 @@ enum class AlbumType {
 data class Album(
     val id: String,
     val name: String,
-    @SerialName("album_type") val albumType: AlbumType? = null,
+    @SerialName("album_type") val albumType: AlbumType,
     @SerialName("total_tracks") val totalTracks: Int,
     @SerialName("available_markets") val availableMarkets: List<String>? = null,
     @SerialName("external_urls") val externalUrls: Map<String, String>,
@@ -34,7 +34,7 @@ data class PartialAlbum(
     val id: String,
     val name: String,
     val uri: String,
-    @SerialName("album_type") val albumType: AlbumType? = null,
+    @SerialName("album_type") val albumType: AlbumType,
     @SerialName("total_tracks") val totalTracks: Int,
     @SerialName("available_markets") val availableMarkets: List<String>? = null,
     @SerialName("external_urls") val externalUrls: Map<String, String>,

--- a/src/main/kotlin/tv/blademaker/kotify/models/Album.kt
+++ b/src/main/kotlin/tv/blademaker/kotify/models/Album.kt
@@ -2,6 +2,14 @@ package tv.blademaker.kotify.models
 
 import kotlinx.serialization.SerialName
 import kotlinx.serialization.Serializable
+import tv.blademaker.kotify.Kotify
+
+@Serializable
+enum class AlbumType {
+    @SerialName("album") ALBUM,
+    @SerialName("single") SINGLE,
+    @SerialName("compilation") COMPILATION
+}
 
 @Serializable
 data class Album(

--- a/src/main/kotlin/tv/blademaker/kotify/models/PlaylistPagination.kt
+++ b/src/main/kotlin/tv/blademaker/kotify/models/PlaylistPagination.kt
@@ -1,12 +1,13 @@
 package tv.blademaker.kotify.models
 
 import io.ktor.http.*
-import kotlinx.serialization.SerialName
-import kotlinx.serialization.Serializable
+import kotlinx.serialization.*
+import tv.blademaker.kotify.serializers.ItemListSerializer
 
 @Serializable
 data class PlaylistPagination(
     val href: String,
+    @Serializable(with = ItemListSerializer::class)
     val items: List<Item>,
     val limit: Int,
     val next: String? = null,

--- a/src/main/kotlin/tv/blademaker/kotify/serializers/ItemListSerializer.kt
+++ b/src/main/kotlin/tv/blademaker/kotify/serializers/ItemListSerializer.kt
@@ -1,0 +1,47 @@
+package tv.blademaker.kotify.serializers
+
+import kotlinx.serialization.ExperimentalSerializationApi
+import kotlinx.serialization.KSerializer
+import kotlinx.serialization.SerializationException
+import kotlinx.serialization.Serializer
+import kotlinx.serialization.builtins.ListSerializer
+import kotlinx.serialization.encoding.Decoder
+import kotlinx.serialization.encoding.Encoder
+import kotlinx.serialization.json.*
+import tv.blademaker.kotify.Kotify
+import tv.blademaker.kotify.models.PlaylistPagination
+
+@OptIn(ExperimentalSerializationApi::class)
+@Serializer(forClass = PlaylistPagination.Item::class)
+object ItemListSerializer : KSerializer<List<PlaylistPagination.Item>> {
+    private val listSerializer = ListSerializer(PlaylistPagination.Item.serializer())
+
+    /**
+     * Deserialize the entire playlist but do not serialize local tracks
+     * and prevent to throw exception when a local track is serialized but
+     * not includes much of the required fields
+     *
+     * @since 0.4.5
+     */
+    override fun deserialize(decoder: Decoder): List<PlaylistPagination.Item> {
+        check(decoder is JsonDecoder) {
+            "This decoder is only supported with Json"
+        }
+
+        return decoder.decodeJsonElement().jsonArray.mapNotNull {
+            try {
+                Kotify.JSON.decodeFromJsonElement(PlaylistPagination.Item.serializer(), it).takeIf { i -> !i.isLocal }
+            } catch (e: SerializationException) {
+                null
+            }
+        }
+    }
+
+    override fun serialize(encoder: Encoder, value: List<PlaylistPagination.Item>) {
+        check(encoder is JsonEncoder) {
+            "This encoder is only supported with Json"
+        }
+
+        listSerializer.serialize(encoder, value)
+    }
+}

